pallets/otc-settlements/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+					asset_out: asset_a,
+				}),
+			Error::<T>::InvalidRoute
+		);
+
+		// get initial otc and router price
+		let otc_price =
+			FixedU128::checked_from_rational(otc.amount_out, otc.amount_in).ok_or(ArithmeticError::Overflow)?;
+
+		// Router trade is disabled in the benchmarks, so disable this one as well.
+		// Without disabling it, the requirements for the extrinsic cannot be met (e.g. profit).
+		#[cfg(not(feature = "runtime-benchmarks"))]
+		if otc.partially_fillable && amount != otc.amount_in {
+			log::debug!(
+			target: "offchain_worker::settle_otc",
+				"calling partial fill order: amount {:?} ", amount);
+			pallet_otc::Pallet::<T>::partial_fill_order(RawOrigin::Signed(pallet_acc.clone()).into(), otc_id, amount)?;
+		} else {
+			log::debug!(
+			target: "offchain_worker::settle_otc",
+				"calling fill order");
+			pallet_otc::Pallet::<T>::fill_order(RawOrigin::Signed(pallet_acc.clone()).into(), otc_id)?;
+		};
+
+		let otc_amount_out = <T as Config>::Currency::free_balance(asset_b, &pallet_acc)
+			.checked_sub(asset_b_balance_before)
+			.unwrap();
+
+		log::debug!(
+			target: "offchain_worker::settle_otc",
+			"calling router sell: amount_in {:?} ", otc_amount_out);
+
+		// Disable in the benchmarks and use existing weight from the router pallet.
+		#[cfg(not(feature = "runtime-benchmarks"))]
+		T::Router::sell(
+			RawOrigin::Signed(pallet_acc.clone()).into(),
+			asset_b,
+			asset_a,
+			otc_amount_out,
+			1,
+			route.clone(),
+		)
+		.map_err(|_| Error::<T>::TradeAmountTooHigh)?;
+		// There are 3 possible types of error:
+		// min trade limit not reached - we start with the largest possible amount, so we can't increase it more.
+		// max trade limit reached - we are interested in this one. We can decrease the amount and try again.
+		// some other error - we can't handle this one properly.
+
+		// // Compare OTC and Router price
+		let router_price_after = T::Router::spot_price_with_fee(&route).unwrap();
+		log::debug!(
+			target: "offchain_worker::settle_otc",
+			"final router price: {:?}   otc_price: {:?} ",
+			router_price_after,
+			otc_price
+		);
+
+		// in the case of fully fillable orders, the resulting price is not important
+		if otc.partially_fillable {
+			let price_diff = {
+				if otc_price > router_price_after {
+					otc_price.saturating_sub(router_price_after)
+				} else {
+					router_price_after.saturating_sub(otc_price)
+				}
+			};
+
+			let price_precision = T::PricePrecision::get()
+				.checked_mul(&otc_price)
+				.ok_or(ArithmeticError::Overflow)?;
+			if price_diff > price_precision {
+				ensure!(router_price_after <= otc_price, Error::<T>::TradeAmountTooHigh);
+				ensure!(router_price_after >= otc_price, Error::<T>::TradeAmountTooLow);
+			}
+		}
+
+		let asset_a_balance_after_router_trade = <T as Config>::Currency::free_balance(asset_a, &pallet_acc);
+
+		let profit = asset_a_balance_after_router_trade
+			.checked_sub(amount)
+			.ok_or(ArithmeticError::Overflow)?;
```

Comment:
nitpick: underflow.

Reply:
you can have integer underflow. it is used eg in balances pallet too. but fine with me.